File: solve.js
var allThreeLetterCombos = [];
var alphebet = "abcdefghijklmnopqrstuvwxyz";

for(var i=0;i<26;i++) {
   for (var j=0;j<26;j++) {
      for(var k=0;k<26;k++) {
	 allThreeLetterCombos.push(alphebet[i] + alphebet[j] + alphebet[k]);
      }
   }
}

var file = require('fs').readFileSync('/usr/share/dict/words', 'utf8');
var lines = file.split("\n");
var output = "";
for(var i=0;i<lines.length;i++){
   lines[i] = lines[i].toLowerCase();
}

for(var i=0;i<allThreeLetterCombos.length;i++) {
   var shortest = null;
   for(var j=0;j<lines.length;j++) {
      if(areInOrder(allThreeLetterCombos[i], lines[j])) {
	 if(shortest == null || shortest.length > lines[j].length) {
	    shortest = lines[j];
	 }
      }
   }
   if(shortest != null) output += allThreeLetterCombos[i] + " - " + shortest + "\n";
   else console.log("" + allThreeLetterCombos[i]);
}

require('fs').writeFileSync('out.txt', output);

function areInOrder(three, word) {
   var ndx = 0;
   for(var i=0;i<word.length;i++) {
      if(three[ndx] == word[i]) {
	 ndx++;
      }
      if(ndx == 3) return true;
   }
   return false;
}

File: solve.rb
words = File.read('/usr/share/dict/words').split("\n");

longest_word = words.reduce{|x,y| x.length < y.length ? y : x}
impossibly_long_word = longest_word + 'x'

license_plates = ('a'..'z').to_a.repeated_permutation(3).map(&:join)
shortest_plates = Hash[license_plates.collect{|p| [p, impossibly_long_word]}]


words.each do |word|
  word.chars.combination(3).each do |plate|
    plate = plate.join.downcase
    shortest_plates[plate] = word if word.length < shortest_plates[plate].length
  end
end


impossible_plates = shortest_plates.select{|k,v| v == impossibly_long_word}

shortest_plates.reject!{|k,v| v == impossibly_long_word}

File.write("impossible_plates.txt", impossible_plates.keys.sort.join("\n"))
File.write("plate_solutions.txt", shortest_plates.map{|k,v| "#{k} => #{v}"}.sort.join("\n"))
